TCH fixes remove quotation marks from `Literal` strings

The fixes for TCH001, TCH002, and TCH003 sometimes delete quotation marks from `Literal`s when `lint.flake8-type-checking.quote-annotations = true`. This can make a program fail to type-check. I am not sure exactly what makes it delete quotation marks but it seems to only happen to literals within nested brackets.

```console
$ ruff --version
ruff 0.7.4

$ cat tch003.py
from collections.abc import Sequence
from typing import Callable, Literal
def f(x: Sequence[Callable[[Literal["1"]], bool]]) -> bool:
    return False
def g(x: Literal["1"]) -> bool:
    return False
print(f([g]))

$ mypy tch003.py
Success: no issues found in 1 source file

$ ruff check --isolated --config 'lint.flake8-type-checking.quote-annotations = true' --select TCH003 tch003.py --unsafe-fixes --fix
Found 1 error (1 fixed, 0 remaining).

$ cat tch003.py
from typing import Callable, Literal, TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Sequence
def f(x: "Sequence[Callable[[Literal[1]], bool]]") -> bool:
    return False
def g(x: Literal["1"]) -> bool:
    return False
print(f([g]))

$ mypy tch003.py
tch003.py:9: error: List item 0 has incompatible type "Callable[[Literal['1']], bool]"; expected "Callable[[Literal[1]], bool]"  [list-item]
Found 1 error in 1 file (checked 1 source file)
```

With `"x"` instead of `"1"` in that example, `"x"` is rewritten to `x` and the final mypy error is:
```console
$ mypy tch003.py
tch003.py:5: error: Name "x" is not defined  [name-defined]
Found 1 error in 1 file (checked 1 source file)
```

--- a/crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs
+++ b/crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs
@@ -392,6 +392,24 @@ impl<'a> SourceOrderVisitor<'a> for QuoteAnnotator<'a> {
                 }
                 self.state.pop();
             }
+            Expr::List(ast::ExprList { elts, .. }) => {
+                let Some((first, remaining)) = elts.split_first() else {
+                    return;
+                };
+                self.annotation.push_str("[");
+                self.visit_expr(first);
+                if let Some(last) = self.state.last_mut() {
+                    if *last == QuoteAnnotatorState::AnnotatedFirst {
+                        *last = QuoteAnnotatorState::AnnotatedRest;
+                    }
+                }
+                for expr in remaining {
+                    self.annotation.push_str(", ");
+                    self.visit_expr(expr);
+                }
+                self.annotation.push_str("]");
+                self.state.pop();
+            }
             Expr::BinOp(ast::ExprBinOp {
                 left, op, right, ..
             }) => {

--- a/crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs
+++ b/crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs
@@ -396,7 +396,7 @@ impl<'a> SourceOrderVisitor<'a> for QuoteAnnotator<'a> {
                 let Some((first, remaining)) = elts.split_first() else {
                     return;
                 };
-                self.annotation.push_str("[");
+                self.annotation.push('[');
                 self.visit_expr(first);
                 if let Some(last) = self.state.last_mut() {
                     if *last == QuoteAnnotatorState::AnnotatedFirst {
@@ -407,7 +407,7 @@ impl<'a> SourceOrderVisitor<'a> for QuoteAnnotator<'a> {
                     self.annotation.push_str(", ");
                     self.visit_expr(expr);
                 }
-                self.annotation.push_str("]");
+                self.annotation.push(']');
                 self.state.pop();
             }
             Expr::BinOp(ast::ExprBinOp {

